File: Codigos Examen l prog/C++/ArregloDesordenado.cpp
#include <iostream>
#include <vector> // Usaremos std::vector para un manejo más flexible del tamaño,
                  // pero puedes usar un arreglo C-style si lo prefieres para la limitación.

// Para un arreglo C-style de tamaño fijo:
// const int MAX_SIZE = 10;
// int arreglo[MAX_SIZE];
// int tamActual = 0;

void inicializarArreglo(std::vector<int>& arr) {
    arr.clear(); // Vacía el vector
    std::cout << "Arreglo inicializado." << std::endl;
}

void mostrarArreglo(const std::vector<int>& arr) {
    if (arr.empty()) {
        std::cout << "El arreglo esta vacio." << std::endl;
        return;
    }
    std::cout << "Arreglo: [";
    for (size_t i = 0; i < arr.size(); ++i) {
        std::cout << arr[i];
        if (i < arr.size() - 1) {
            std::cout << ", ";
        }
    }
    std::cout << "]" << std::endl;
}

void insertarElemento(std::vector<int>& arr, int elemento) {
    // Si usaras un arreglo C-style con MAX_SIZE:
    // if (tamActual < MAX_SIZE) {
    //     arreglo[tamActual++] = elemento;
    //     std::cout << "Elemento " << elemento << " insertado." << std::endl;
    // } else {
    //     std::cout << "El arreglo esta lleno, no se puede insertar." << std::endl;
    // }

    arr.push_back(elemento); // std::vector maneja el tamaño dinámicamente
    std::cout << "Elemento " << elemento << " insertado." << std::endl;
}

int buscarElemento(const std::vector<int>& arr, int elemento) {
    for (size_t i = 0; i < arr.size(); ++i) {
        if (arr[i] == elemento) {
            return i; // Devuelve la posicion (0-based)
        }
    }
    return -1; // No encontrado
}

void eliminarElemento(std::vector<int>& arr, int elemento) {
    int pos = buscarElemento(arr, elemento);
    if (pos != -1) {
        arr.erase(arr.begin() + pos); // Elimina el elemento en la posición
        std::cout << "Elemento " << elemento << " eliminado." << std::endl;
    } else {
        std::cout << "Elemento " << elemento << " no encontrado para eliminar." << std::endl;
    }
}

int main() {
    std::vector<int> arreglo; // Usamos std::vector
    int opcion;
    int valor;
    int posicion;

    inicializarArreglo(arreglo);

    do {
        std::cout << "\n--- MENU ARREGLO DESORDENADO ---" << std::endl;
        std::cout << "1. Crear/Reinicializar Arreglo" << std::endl;
        std::cout << "2. Insertar Elemento" << std::endl;
        std::cout << "3. Buscar Elemento" << std::endl;
        std::cout << "4. Eliminar Elemento" << std::endl;
        std::cout << "5. Mostrar Arreglo" << std::endl;
        std::cout << "0. Salir" << std::endl;
        std::cout << "Ingrese su opcion: ";
        std::cin >> opcion;

        switch (opcion) {
            case 1:
                inicializarArreglo(arreglo);
                break;
            case 2:
                std::cout << "Ingrese el valor a insertar: ";
                std::cin >> valor;
                insertarElemento(arreglo, valor);
                break;
            case 3:
                std::cout << "Ingrese el valor a buscar: ";
                std::cin >> valor;
                posicion = buscarElemento(arreglo, valor);
                if (posicion != -1) {
                    std::cout << "Elemento " << valor << " encontrado en la posicion " << posicion << "." << std::endl;
                } else {
                    std::cout << "Elemento " << valor << " no encontrado." << std::endl;
                }
                break;
            case 4:
                std::cout << "Ingrese el valor a eliminar: ";
                std::cin >> valor;
                eliminarElemento(arreglo, valor);
                break;
            case 5:
                mostrarArreglo(arreglo);
                break;
            case 0:
                std::cout << "Saliendo del programa." << std::endl;
                break;
            default:
                std::cout << "Opcion invalida. Intente de nuevo." << std::endl;
        }
    } while (opcion != 0);

    return 0;
}

File: Codigos Examen l prog/Pascal/ArregloDesordenado.pas
program ArregloDesordenado;

const
  MAX_SIZE = 10;

type
  TArreglo = array[1..MAX_SIZE] of Integer;

var
  arreglo: TArreglo;
  tamActual: Integer;
  opcion: Integer;
  valor: Integer;
  posicion: Integer;

procedure InicializarArreglo(var arr: TArreglo; var tam: Integer);
begin
  tam := 0; // El arreglo está vacío
  Writeln('Arreglo inicializado.');
end;

procedure MostrarArreglo(const arr: TArreglo; tam: Integer);
var
  i: Integer;
begin
  if tam = 0 then
  begin
    Writeln('El arreglo esta vacio.');
    exit;
  end;
  Write('Arreglo: [');
  for i := 1 to tam do
  begin
    Write(arr[i]);
    if i < tam then
      Write(', ');
  end;
  Writeln(']');
end;

procedure InsertarElemento(var arr: TArreglo; var tam: Integer; elemento: Integer);
begin
  if tam < MAX_SIZE then
  begin
    inc(tam);
    arr[tam] := elemento;
    Writeln('Elemento ', elemento, ' insertado.');
  end
  else
  begin
    Writeln('El arreglo esta lleno, no se puede insertar.');
  end;
end;

function BuscarElemento(const arr: TArreglo; tam: Integer; elemento: Integer): Integer;
var
  i: Integer;
begin
  BuscarElemento := -1; // No encontrado
  for i := 1 to tam do
  begin
    if arr[i] = elemento then
    begin
      BuscarElemento := i; // Devuelve la posicion (1-based)
      exit;
    end;
  end;
end;

procedure EliminarElemento(var arr: TArreglo; var tam: Integer; elemento: Integer);
var
  pos: Integer;
  i: Integer;
begin
  pos := BuscarElemento(arr, tam, elemento);
  if pos <> -1 then
  begin
    // Mover el ultimo elemento a la posicion del eliminado
    for i := pos to tam - 1 do
    begin
      arr[i] := arr[i + 1];
    end;
    dec(tam);
    Writeln('Elemento ', elemento, ' eliminado.');
  end
  else
  begin
    Writeln('Elemento ', elemento, ' no encontrado para eliminar.');
  end;
end;

begin
  InicializarArreglo(arreglo, tamActual);

  repeat
    Writeln('');
    Writeln('--- MENU ARREGLO DESORDENADO ---');
    Writeln('1. Crear/Reinicializar Arreglo');
    Writeln('2. Insertar Elemento');
    Writeln('3. Buscar Elemento');
    Writeln('4. Eliminar Elemento');
    Writeln('5. Mostrar Arreglo');
    Writeln('0. Salir');
    Write('Ingrese su opcion: ');
    Readln(opcion);

    case opcion of
      1: InicializarArreglo(arreglo, tamActual);
      2:
        begin
          Write('Ingrese el valor a insertar: ');
          Readln(valor);
          InsertarElemento(arreglo, tamActual, valor);
        end;
      3:
        begin
          Write('Ingrese el valor a buscar: ');
          Readln(valor);
          posicion := BuscarElemento(arreglo, tamActual, valor);
          if posicion <> -1 then
            Writeln('Elemento ', valor, ' encontrado en la posicion ', posicion, '.')
          else
            Writeln('Elemento ', valor, ' no encontrado.');
        end;
      4:
        begin
          Write('Ingrese el valor a eliminar: ');
          Readln(valor);
          EliminarElemento(arreglo, tamActual, valor);
        end;
      5: MostrarArreglo(arreglo, tamActual);
      0: Writeln('Saliendo del programa.');
      else Writeln('Opcion invalida. Intente de nuevo.');
    end;
  until opcion = 0;
end.